text/0000-sbom.md

```diff
@@ -0,0 +1,118 @@
+
+# Motivation
+[motivation]: #motivation
+
+## Why should we do this?
+
+Quoting [NTIA Whitepaper][NTIA-WP] - 
+> Modern software systems involve increasingly complex and dynamic supply chains. Unfortunately, the composition and functionality of these systems lacks transparency; this contributes substantially to cybersecurity risks, alongside the cost of development, procurement, and maintenance. This has broad implications in our interconnected world; risk and cost affect collective goods, like public safety and national security, in addition to the products and services upon which businesses rely.
+> We propose that increased supply chain transparency through SBOMs can reduce cybersecurity risks and overall costs by:
+> - Enhancing the identification of vulnerable systems and the root cause of incidents
+> - Reducing unplanned and unproductive work
+> - Supporting more informed market differentiation and component selection
+> - Reducing duplication of effort by standardizing formats across multiple sectors
+> - Identifying suspicious or counterfeit software components
+
+Currently since container images built by Cloud Native Buildpacks don't follow a generic file system layout, they may not be easily scannable by container scanning tools. We do however provide a BOM which should greatly help sidestep the entire manual scanning process and speed up things like CVE detection by directly providing the CVE scanner with a BOM.
+
+However, since we currently do not impose any standard for what the BOM should look like and since the metadata table in BOM is a freeform table, it is very hard to have consistent BOMs that can be used by CVE scanners.
+
+Having a structured and consistent SBOM that can be consumed by downstream systems could greatly enhance the security proposition around buildpacks.
+
+## What use cases does it support?
+
+- Consistent SBOM across buildpacks
+- Improved integration with other supply chain analysis and scanning tools
+
+## What is the expected outcome?
+
+- Use of CycloneDX as the SBOM storage format and updates the the specification and lifecycle to reflect the same.
+- Moving the SBOM from a label to a file in a separate layer.
+- Enhancements to the platform spec and `pack` to allow easy interaction with SBOMs including conversion between major SBOM formats.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Currently the SBOM is defined in the `<layer>.toml`, `launch.toml` and `build.toml` respectively under the `[bom]` table.
+It may not be the most user-friendly way for buildpack authors to create SBOM documents in the above format. This RFC proposes that the `bom` be moved to `<layer>.bom.<ext>`, `launch.bom.<ext>` and `build.bom.<ext>` instead where `<ext>` may be `xml` or `json`. This is mostly to allow for different BOM generation tools to be used. This also allows us in the future to add support for `spdx`  tag documents or other file formats if needed.
```

Comment:
>where `<ext>` may be `xml` or `json`

It seems that having support for multiple data representations would put extra burden on the lifecycle to understand all of those, convert, merge, and ultimately pull them into one representation in the CycloneDX format.

What about picking one standard format like "json", and pushing the formatting/conversion responsibilities to the integrator/buildpack (`libcnb` can probably help some with minimizing the work on an author too).

For example, if I want to use tool XYZ that only outputs in XML format. I as the buildpack author would be responsible for running the tool to get the SBOM and then converting it to JSON. The benefit of this, IMHO, would be that tool format/conversion issues would be on the user of the tool to fix, not on the lifecycle team.

Reply:
100% this

Reply:
FYI CycloneDX and SPDX tooling already exists to convert between JSON and XML representations.

Reply:
This makes sense to me. I guess JSON might be an easier target give that our current SBOM label is also json. I guess we can adopt a naming scheme as suggested in the NTIA document around 

`sbom.spdx.json` or `sbom.cdx.json` (I believe the cdx one was a suggesting but not a standard?)

Reply:
Tern uses an internal data model to represent all the keys and then uses plugins to generate each of the formats based on a internal-model-key to format-key template. Hence there is no data loss when generating each format. The loss happens when reading the formats for obvious reasons.